rtstats/rtstats.1.in
.TH RTSTATS 1 "2010-03-10"
.SH NAME
rtstats - Generate time based statistics from an LDM product queue and transmit
to a receiving host using LDM protocol.
.SH SYNOPSIS
.HP
.ft B
rtstats
.nh
\%[-v]
\%[-x]
\%[-H\ \fIhostname\fP]
\%[-h\ \fIremote host\fP]
\%[-P\ \fIport\fP]
\%[-l\ \fIlogdest\fP]
\%[-f\ \fIfeedtype\fP]
\%[-p\ \fIpattern\fP]
\%[-q\ \fIqueue\fP]
\%[-o\ \fIoffset\fP]
\%[-i\ \fIinterval\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This program scans an LDM product queue (see \fBpq\fP(3)) and creates 
products containing statistics about products being received.
The products are generated for each incoming feedtype and are assigned 
the EXP feedtype for transmission. Statistics are reported at 1 minute
intervals as new data are being received.

.B Rtstats
is typically used in to accumlate data for monitoring the performance
of a distributed LDM system. Statistics for various nodes in the system can 
be compared to ascertain the performance of the network. This program is
similar to \fBpqbinstats\fP(1) with the addition of the delivery of
the statistics via LDM protocol to a specified host.

A separate statistic product will be sent for each feedtype and LDM product origin.
That is, each LDM product will be named as follows:

.RS +4
  rtstats-LDMVERSION/LDMHOST/FEEDTYPE/PRODUCT_ORIGIN
.RE

where \fILDMVERSION\fP is the release number of the LDM distribution,
\fILDMHOST\fP is the machine where \fBrtstats\fP(1) is running, \fIFEEDTYPE\fP
is the class of the incoming data, and \fIPRODUCT_ORIGIN\fP is the assigned
identifier denoting the original injection host of the data
and upstream host identifier if available (LDM 5.2 and later).
In addition to cumulative statistics for the individual feeds,
the instantaneous product latency for the most recent product received 
is provided. The arriving statistics may be processed through
\fBpqact\fP(1).

.SH REPORTED STATISTICS
.TP
Creation-time of most recent product
.TP
Product-queue insertion-time of most recent product
.TP
Receiving host (eg localhost id)
.TP
Feedtype
.TP
Origin host
.TP
Products received since \fBrtstats\fP(1) started
.TP
Bytes received since \fBrtstats\fP(1) started
.TP
Latency of most recent product
.TP
Average latency of products received since \fBrtstats\fP(1) started
.TP
Peak latency@min/sec past the hour since \fBrtstats\fP(1) started
.TP
Version of LDM running on localhost

.SH OPTIONS
.TP
.B -v
Verbose logging.  A line is emitted for every product in the queue whose
feed type and identifier match the specified feedtype and pattern.  The
emitted line contains the UTC date and time, the program name, the product
ingest time, feedtype, sequence number, size in bytes, and the product
identifier.
.TP
.B -x
Debugging information is also emitted.
.TP
.BI "-H " hostname
Report as host \fIhostname\fP. The default is to report as the value of
registry parameter \fBregpath{HOSTNAME}\fP, if it is set, or the return value
of the \fBgethostname\fP(3) function, otherwise.
.TP
.BI "-h " remote
Requests that the statistic products be delivered to the LDM  server on
\fIremote\fP.
Defaults to localhost if not specified.
.TP
.BI \-l " logdest"
Log to \fIdest\fP. One of \fB''\fP (system logging daemon), \fB'-'\fP
(standard error stream), or file \fIlogdest\fP. Default is the standard error
stream if the process has a controlling terminal (i.e., the process isn't a
daemon); otherwise, either the LDM log file or the system logging daemon
(execute this program with just the option \fB'-?'\fP to determine which).
.TP
.BI \-f " feedtype"
Reads from the product queue only products that have a feedtype that is a
member of the \fIfeedtype\fP set.  The default is "\fBANY\fP", specifying all
feed types.  See \fBpqact\fP(1) for a list of basic feed types
and the use of feedtype expressions to specify combinations of basic
feed types.
.TP
.BI \-p " pattern"
Reads from the product queue only products whose identifier
matches the regular expression \fIpattern\fP.
The default is "\fB.*\fP", specifying all products.
.TP
.BI "-q " queue
The pathname of the data product queue.
The default is
.nh
\fB$(regutil regpath{QUEUE_PATH})\fP.
.hy
The configuration default can
be overridden by setting the environment variable \fBLDMPQFNAME\fP.
.TP
.BI \-i " interval"
Polling interval, in seconds.  When the end of the queue is reached, the
program sleeps and
checks for new products in the product queue every \fIinterval\fP seconds.
When
\fBrtstats\fP(1)
is run in the same process group as the programs that insert products into
the product queue, a signal informs
\fBrtstats\fP(1)
and all other interested processes in the process group
whenever a new product is available.
In this case, the interval should be set so that polling timer rarely expires.
The default interval is 15 seconds.
.TP
.BI \-o " offset"
Offset time, in seconds.
Begin reading products inserted into the product queue \fIoffset\fP
seconds earlier than the current time.
The default is to begin just after the time stamp of the oldest product
in the queue at startup.
.TP
.BI \-P " port"
Port on the remote system to which to connect.  The well-known LDM port is
388.
.SH SIGNALS
.TP
.BR SIGTERM
Normal termination.
.TP
.B SIGUSR1
Refresh logging (if configure(1) executed without the "--with-ulog" option).
.TP
.B SIGUSR2
Toggle logging of verbose messages.

.SH EXAMPLE
The following line may be added to the ldmd.conf file to automatically
start \fBrtstats\fP(1) when the LDM is started and deliver statistics to
the collection server at the Unidata Program Center.
.RS +4
  exec    "rtstats -h rtstats.unidata.ucar.edu"
.RE

.SH "SEE ALSO"
.LP
.BR ldmd (1),
.BR pqact (1),
.BR pqcat (1),
.BR pqbinstats (1),
.BR ulog (3),
.BR pq (3),
.BR gethostbyname (3),
.BR syslogd (8),
\fBhttp://www.unidata.ucar.edu/software/ldm/\fP.

.SH SUPPORT
.LP
If you have problems with this program, then you should first examine the 
LDM email archive for similar problems and how they were solved.
The email archive is available via the following World Wide Web URL:
.sp
.RS
\fBhttp://www.unidata.ucar.edu/software/ldm/\fP
.RE
.sp
If this does not suffice and your site is a member of the Unidata 
program, then send an inquiry via email -- together will all relevant 
information -- to
.sp
.RS
\fBsupport@unidata.ucar.edu\fP
.RE


CHANGE_LOG
6.14.5.2
    Documentation:
        * Removed empty entry from list of general platform requirements

6.14.5.1  2023-01-31T16:15:39-0700
    rtstats(1):
        * Commented code about what function logs messages because the code is convoluted
        * Removed trailing newline in rtstats(1) log message to eliminate extraneous log message
        * Eliminated sending of NULLPROC before closing connection because it's superfluous
        * Documented return value of ldmsend_statsbin()
        * Eliminated possibility of overrun in formatting report
        * Eliminated unnecessary log message
        * Replaced abort() due to unsupported remote LDM version with log_error() and ECONNABORTED
          return value
        * man(1) page: Corrected description of number of bytes, number of products, and
          peak-latency parameters: they are not for "this hour" but since rtstats(1) was started

    Documentation:
        * Added requirement that "RemoveIPC" in "/etc/systemd/logind.conf" must be "no" to
          prevent systemd(8) from deleting the LDM's shared-memory segment that contains the
          upstream LDM database.

    CHANGE_LOG:
        * Consolidated 6.14.4.* entries into 6.14.4. Should have done that earlier.

6.14.4  2023-01-26T11:03:17-0700
    ldmd(1):
        * Eliminated log messages when an upstream LDM process that was ACCEPTed terminates. This
          should reduce the number of log messages on the rtstats(1) mothership.
        * Moved setting of "hiyaCalled" to top of HIYA service functions
        * Made ECONNRESET error a single INFO-level log message
        * Replace some ERROR log messages with FATAL ones in the main module.

    NOAAPort:
        * Added entries with new PDTNs to GEMPAK tables:
            * g2varswmo.tbl:
                000 002 000 002 Wind Direction             Degree true  DRCT    0  -9999.00
            * g2varsncep.tbl:
                001 001 194 009 Prob precip exceeding FFG  %            PEFG--  0  -9999.00

    rtstats(1):
        * Made error reporting more robust

    Documentation:
        * More hyperlink corrections
        * Returned Coverity Scan badge to top-level, version-specific index.html

    Misc:
        * Removed lint found by Coverity Scan
        * ErrObj module
            * Created
            * Created log_dispose(ErrObj*)
            * Might get used; might not.

    DevOps:
        * Enhanced "pre-commit" hook to vet the version ID in CHANGE_LOG and ensure a timestamp
          in CHANGE_LOG when committing to the main branch
        * Enhanced "OnPush.yml":
            * Vet the version ID in CHANGE_LOG and "configure.ac"
            * Update the "versions.html" file for LDM documentation on the artifacts server
        * Deleted "--disable-root-actions" from Coverity Scan target
        * Added deletion of /tmp/pqact_test.conf before creation due to /tmp having the "restricted
          deletion" bit set
        * Deleted use of CodeQL in OnPush.yml in favor of using Coverity Scan because it's better,
          IMO

6.14.3  2022-12-22T13:34:03-0700
    Documentation:
        * Fixed more links in HTML documentation

6.14.2	2022-12-20T15:32:36-0700
    Documentation:
        * Added missing graphic "product-queue.svg"
        * Fixed links in HTML documentation

6.14.1  2022-12-19T12:40:42-0700
    Installation:
        * Eliminated the configure(1)-script option "--disable-root-actions". Installation now
          proceeds as if this option were always specified. The "make install" step now prints a
          message at the end stating that the superuser (i.e., "root") must now execute the command
          "make root-actions". This was necessary in order to migrate the continuous-integration
          workflow to GitHub actions: autoconf(1) 2.71 on the Ubuntu 22.04 platform creates an
          invalid configure(1) script if the autoconf(1) configuration-file contains this option.

          Installation from the unpacked distribution now comprises the following steps (details
          omitted):
            - Execute the configure(1) script as the LDM user
            - "make install" as the LDM user
            - "make root-actions" as root
          See the HTML documentation for details.
        * The configure(1) script now requires that a C++ compiler be found if the "--with-noaaport"
          option is specified

    ldmd(1):
        * Demoted the logging level of some common connection losses to reduce the number of ERROR
          messages in the log file
        * Improved log message when select() fails
        * Improved logging and simplified up6_run()
        * Only close file-descriptors 0, 1, & 2 when daemonizing
        * Added "-h" (help) option
        * Moved daemonization to before the creation of the shared upstream LDM database
        * Ensured that the LDM server is the process group leader.

    regutil(1):
        * Added "-l <logdest>" option to accomodate GitHub actions

    pqing(1):
        * Eliminated compilation error in "tty.c" under MacOS by including <sys/ioctl.h>

    NOAAPort:
        * blender(1):
            * Added frame-header sequence-number to determination of frame ordering.
            * Corrected determination of frame ordering
        * GEMPAK Tables: Added missing PDTNs

    LDM registry:
        * Ensured that the file-descriptor of the registry is close-on-exec

    Logging:
        * In order to accomodate GitHub actions, changed the determination of whether or not to use
          the standard error stream from whether of not the process was a daemon to whether or not
          the standard error stream is open on /dev/null

    rpc(3) library:
        * Eliminated compilation errors under MacOS by replacing bzero(), bset(), and bcmp() with
          POSIX equivalents
        * Demoted the logging level of some common connection losses to reduce the number of
          ERROR messages in the log file

    Documentation:

rtstats/rtstats.c
/**
 * Reports performance statistics on the product-queue to an LDM server.
 *
 * Copyright 2018, University Corporation for Atmospheric Research
 * All rights reserved. See file COPYRIGHT in the top-level source-directory for
 * copying and redistribution conditions.
 */

#include <config.h>
#include <libgen.h>
#include <fcntl.h>
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <regex.h>
#include "ldm.h"
#include "atofeedt.h"
#include "globals.h"
#include "inetutil.h"
#include "remote.h"
#include "ldmprint.h"
#include "log.h"
#include "pq.h"
#ifndef HAVE_SETENV
    #include "setenv.h"
#endif
#include "RegularExpressions.h"

#ifdef NO_ATEXIT
#include "atexit.h"
#endif

#ifndef DEFAULT_INTERVAL
#define DEFAULT_INTERVAL 15
#endif

static volatile sig_atomic_t intr = 0;
static volatile sig_atomic_t stats_req = 0;

#ifndef DEFAULT_FEEDTYPE
#define DEFAULT_FEEDTYPE ANY
#endif

/* binstats.c */
extern int binstats(const prod_info *infop,
        const struct timeval *reftimep);

extern void dump_statsbins(void);
extern void syncbinstats(const char* hostname);

unsigned remotePort = LDM_PORT;


/*
 */
/*ARGSUSED*/
static int
addtostats(const prod_info *infop, const void *datap,
                void *xprod, size_t size,  void *notused)
{
        struct timeval tv;
        pq_ctimestamp(pq, &tv);
        if(tvIsNone(tv))
                tv = TS_ZERO;
        if(log_is_enabled_info)
                log_info_q("%s", s_prod_info(NULL, 0, infop,
                        log_is_enabled_debug));
        binstats(infop, &tv);
        return 0;
}


static void
usage(const char *av0) /*  id string */
{
    log_error_q(
"Usage: %s [options]", av0);
    log_error_q(
"where:");
    log_error_q(
"    -v           Log INFO-level messages (log each product).");
    log_error_q(
"    -x           Log DEBUG-level messages.");
    log_error_q(
"    -l dest      Log to `dest`. One of: \"\" (system logging daemon), \"-\"");
    log_error_q(
"                 (standard error), or file `dest`. Default is \"%s\"",
            log_get_default_destination());
    /*
     * NB: Don't use "s_feedtypet(DEFAULT_FEEDTYPE)" in the following because
     * it looks ugly for "ANY - EXP".
     */
    log_error_q(
"    -f feedtype  Scan for data of type \"feedtype\" (default: ");
    log_error_q(
"                 \"ANY - EXP\").");
    log_error_q(
"    -p pattern   Interested in products matching \"pattern\"");
    log_error_q(
"                 (default: \".*\").") ;
    log_error_q(
"    -q queue     Use file \"queue\" as product-queue (default: ");
    log_error_q(
"                 \"%s\").", getDefaultQueuePath());
    log_error_q(
"    -o offset    Oldest product to consider is \"offset\"");
    log_error_q(
"                 seconds before now (default: 0).");
    log_error_q(
"    -i interval  Poll queue every \"interval\" seconds (default:");
    log_error_q(
"                 %d).", DEFAULT_INTERVAL);
    log_error_q(
"    -h hostname  Send to LDM server on host \"hostname\"");
    log_error_q(
"                 (default: LOCALHOST).");
    log_error_q(
"    -P port      Send to port \"port\" (default: %d).", LDM_PORT);
    exit(1);
}


void
cleanup(void)
{
        log_notice_q("Exiting");

        if(pq && !intr) 
                (void)pq_close(pq);

        log_fini();
}


static void
signal_handler(int sig)
{
    switch(sig) {
    case SIGINT :
        intr = 1;
        exit(0);
        /*NOTREACHED*/
    case SIGTERM :
        done = 1;
        return;
    case SIGUSR1 :
        log_refresh();
        stats_req = 1;
        return;
    case SIGUSR2 :
        log_roll_level();
        return;
    }
}


/*
 * register the signal_handler
 */
static void
set_sigactions(void)
{
    struct sigaction sigact;
    sigemptyset(&sigact.sa_mask);
    sigact.sa_flags = 0;

    /* Ignore the following */
    sigact.sa_handler = SIG_IGN;
    (void) sigaction(SIGPIPE, &sigact, NULL);
    (void) sigaction(SIGALRM, &sigact, NULL);
    (void) sigaction(SIGCHLD, &sigact, NULL);

    /* Handle the following */
    sigact.sa_handler = signal_handler;

    /* Don't restart the following */
    (void) sigaction(SIGINT, &sigact, NULL);

    /* Restart the following */
    sigact.sa_flags |= SA_RESTART;
    (void) sigaction(SIGTERM, &sigact, NULL);
    (void) sigaction(SIGUSR1, &sigact, NULL);
    (void) sigaction(SIGUSR2, &sigact, NULL);

    sigset_t sigset;
    (void)sigemptyset(&sigset);
    (void)sigaddset(&sigset, SIGPIPE);
    (void)sigaddset(&sigset, SIGALRM);
    (void)sigaddset(&sigset, SIGCHLD);
    (void)sigaddset(&sigset, SIGTERM);
    (void)sigaddset(&sigset, SIGUSR1);
    (void)sigaddset(&sigset, SIGUSR2);
    (void)sigaddset(&sigset, SIGINT);
    (void)sigprocmask(SIG_UNBLOCK, &sigset, NULL);
}

int main(int ac, char *av[])
{
        const char* const  progname = basename(av[0]);
        prod_class_t       clss;
        prod_spec          spec;
        int                status = 0;
        int                interval = DEFAULT_INTERVAL;
        int                toffset = TOFFSET_NONE;
        extern const char* remote;
        char*              hostname = ghostname();

        /*
         * Setup default logging before anything else.
         */
        if (log_init(av[0])) {
            log_syserr("Couldn't initialize logging module");
            exit(1);
        }

        remote = "localhost";

        if(set_timestamp(&clss.from) != ENOERR) /* corrected by toffset below */
        {
                int errnum = errno;
                fprintf(stderr, "Couldn't set timestamp: %s", 
                        strerror(errnum));
                exit(1);
        }
        clss.to = TS_ENDT;
        clss.psa.psa_len = 1;
        clss.psa.psa_val = &spec;
        spec.feedtype = DEFAULT_FEEDTYPE;
        spec.pattern = ".*";
        
        /*
         * Sigh, in order to read back from existing stats files,
         * we call mktime(3). Since the stats are in UTC,
         * and mktime uses "local" time, the local time for
         * this program must be UTC.
         */
        if(setenv("TZ", "UTC0",1))
        {
                int errnum = errno;
                log_error_q("setenv: Couldn't set TZ: %s", strerror(errnum));
                exit(1);        
        }

        {
            int ch;

            opterr = 0; /* stops getopt() from printing to stderr */

            while ((ch = getopt(ac, av, ":vxl:p:f:q:o:i:H:h:P:")) != EOF) {
                switch (ch) {
                    case 'v':
                        if (!log_is_enabled_info)
                            (void)log_set_level(LOG_LEVEL_INFO);
                        break;
                    case 'x':
                        (void)log_set_level(LOG_LEVEL_DEBUG);
                        break;
                    case 'l':
                        if (log_set_destination(optarg)) {
                            log_syserr("Couldn't set logging destination to \"%s\"",
                                    optarg);
                            usage(progname);
                        }
                        break;
                    case 'H':
                        hostname = optarg;
                        break;
                    case 'h':
                        remote = optarg;
                        break;
                    case 'p':
                        spec.pattern = optarg;
                        if (re_isPathological(spec.pattern)) {
                            log_notice_q("Adjusting pathological regular-expression \"%s\"",
                                    spec.pattern);
                            re_vetSpec(spec.pattern);
                        }
                        break;
                    case 'f': {
                        int fterr = strfeedtypet(optarg, &spec.feedtype) ;
                        if (fterr != FEEDTYPE_OK) {
                            log_error_q("Bad feedtype \"%s\", %s", optarg,
                                    strfeederr(fterr)) ;
                            usage(progname);
                        }
                        break;
                    }
                    case 'q':
                        setQueuePath(optarg);
                        break;
                    case 'o':
                        toffset = atoi(optarg);
                        if(toffset == 0 && *optarg != '0') {
                            log_error_q("Invalid offset %s", optarg);
                            usage(progname);
                        }
                        break;
                    case 'P': {
                        char*       suffix = "";
                        long        port;

                        errno = 0;
                        port = strtol(optarg, &suffix, 0);

                        if (0 != errno || 0 != *suffix ||
                            0 >= port || 0xffff < port) {

                            log_error_q("Invalid port %s", optarg);
                            usage(progname);
                        }

                        remotePort = (unsigned)port;

                        break;
                    }
                    case 'i':
                        interval = atoi(optarg);
                        if (interval == 0 && *optarg != '0') {
                            log_error_q("Invalid interval %s", optarg);
                            usage(progname);
                        }
                        break;
                    case '?':
                        log_error_q("Invalid option \"%c\"", optopt);
                        usage(progname);
                        break;
                    case ':':
                        log_error_q("No argument for option \"%c\"", optopt);
                        usage(progname);
                        break;
                }
            } /* getopt() loop */

            if (optind != ac) {
                log_error_q("Invalid operand: \"%s\"", av[optind]);
                usage(progname);
            }
        } /* command-line decoding block */

        const char* const  pqfname = getQueuePath();

        if (regcomp(&spec.rgx, spec.pattern, REG_EXTENDED|REG_NOSUB)) {
            log_error_q("Bad regular expression \"%s\"\n", spec.pattern);
            usage(progname);
        }

        log_notice_q("Starting Up (%d)", getpgrp());

        /*
         * register exit handler
         */
        if(atexit(cleanup) != 0)
        {
                log_syserr("atexit");
                exit(1);
        }

        /*
         * set up signal handlers
         */
        set_sigactions();


        /*
         * Open the product queue
         */
        status = pq_open(pqfname, PQ_READONLY, &pq);
        if(status)
        {
                if (PQ_CORRUPT == status) {
                    log_error_q("The product-queue \"%s\" is inconsistent\n",
                            pqfname);
                }
                else {
                    log_error_q("pq_open failed: %s: %s\n",
                            pqfname, strerror(status));
                }
                exit(1);
        }

        if(toffset == TOFFSET_NONE)
        {
                /*
                 * Be permissive with the time filter,
                 * jump now to the end of the queue.
                 */
                clss.from = TS_ZERO;
                (void) pq_last(pq, &clss, NULL);
        }
        else
        {
                /*
                 * Filter and queue position set by
                 * toffset.
                 */
                clss.from.tv_sec -= toffset;
                pq_cset(pq, &clss.from);
        }

        while(exitIfDone(0))
        {
                if(stats_req)
                {
                        dump_statsbins();
                        stats_req = 0;
                }

                status = pq_sequence(pq, TV_GT, &clss, addtostats, 0);

                switch(status) {
                case 0: /* no error */
                        continue; /* N.B., other cases sleep */
                case PQUEUE_END:
                        log_debug("End of Queue");
                        break;
                case EAGAIN:
                case EACCES:
                        log_debug("Hit a lock");
                        break;
                default:
                	if (status > 0) {
                        log_add("pq_sequence failed: %s (errno = %d)",
                                strerror(status), status);
                        log_flush_error();
                	}
					exit(1);
					break;
                }

                syncbinstats(hostname);

                if(interval == 0)
                {
                        done = 1;
                        break;
                }

                pq_suspend(interval);
        }

        return 0;
}
